plex/player.py

```diff
@@ -0,0 +1,186 @@
+
+    def onPlayBackSeek(self, time, seekOffset):
+        '''Triggered when seeking.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackSeekChapter(self, chapter):
+        '''Triggered when seeking chapters.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackPaused(self):
+        '''Triggered when playback is paused.'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PAUSED)
+
+    def onPlayBackResumed(self):
+        '''Triggered when playback is resumed after a pause'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackStopped(self):
+        '''Triggered when playback is stopped'''
+        self._syncPlaybackState(state=PLEX_PLAYER_STOPPED)
+        self.onPlayBackEnded()
+
+    def onPlayBackEnded(self):
+        '''Triggered when playback ends. Resets the player state and inherently kills the reporting loop'''
+        self._reset()
+
+    def _startPlayback(self):
+        '''Identifies the item (if from Plex) and initializes the player state'''
+        if not self._file:
+            return
+
+        if not self.isPlayingVideo():
+            return
+
+        videoInfoTag = self.getVideoInfoTag()
+        if not videoInfoTag:
+            return
+
+        itemId = videoInfoTag.getUniqueID(PLEX_PROTOCOL)
+        if not itemId:
+            return
+
+        if not itemId.isdigit():
+            log('invalid item id {} (non digit). Kodi will not report playback state to Plex Media Server' \
+                .format(matchingItems[0]), xbmc.LOGERROR)
+
+        self._itemId = int(itemId)
+
+        for mediaProvider in self._providers.values():
+            importedItems = xbmcmediaimport.getImportedItemsByProvider(mediaProvider)
+            matchingItems = [ importedItem for importedItem in importedItems \
+                if importedItem.getVideoInfoTag() and importedItem.getVideoInfoTag().getUniqueID(PLEX_PROTOCOL) == str(self._itemId) ]
+            if not matchingItems:
+                continue
+
+            if len(matchingItems) > 1:
+                log('multiple items imported from {} match the imported Plex item {} playing from {}' \
+                    .format(mediaProvider2str(mediaProvider), self._itemId, self._file), xbmc.LOGWARNING)
+
+            self._mediaProvider = mediaProvider
+            break
+
+        if self._mediaProvider:
+            # save item
+            plexServer = Server(self._mediaProvider)
+            self._item = Api.getPlexItemDetails(
+                plexServer.PlexServer(),
+                self._itemId,
+                Api.getPlexMediaClassFromMediaType(videoInfoTag.getMediaType())
+            )
+            self._duration = int(self.getTotalTime() * 1000)
+
+            # start playback monitoring thread
+            if not self._playback_monitor or not self._playback_monitor.isAlive():
```

Comment:
Any reason why you don't use the `Progress()` approach used in `mediaimporter.emby`? The player is already running as part of a python service which is constantly running.

Reply:
sounds reasonable

Reply:
To be honest I didn't get the `Process` logic at first. If the observer is a running thread it doesn't make any sense to create a new one and we can simply reuse it to call into the player instance. 👍 

Thinking about it. Is the implementation in mediaimporter.emby thread-safe? I am not 100% sure it is. Player events, AFAIK will be trigged by the PythonInvoker while your observer thread is also updating the same variables. It's probably better to keep a lock on the report playback method (or around any critical member variables). In python you can use contextlib do do it easily:

```python
import threading
self._lock = threading.Lock()

with self._lock:
     do_something()
```

Reply:
You're right. TBH I didn't give it much thought. Adding a lock around the callback actions and the `Process()` logic makes sense.

Reply:
I've created a bug at https://github.com/Montellese/mediaimporter.emby/issues/7 for `mediaimporter.emby`.